### PR TITLE
[2.7] bpo-30258: regrtest handles child process crash

### PR DESCRIPTION
**Pull request by vstinner**
Backport the CHILD_ERROR status from master: a test is considered as
failed if a worker process running a test exited with a code
different than zero.

Change also the output: write stdout and stderr of the child process
after the test name, instead of writing it before.

### CODE REVIEW

File: Lib/test/regrtest.py [hunk 1]
```diff
@@ -478,6 +480,9 @@ def accumulate_result(test, result):
         elif ok == RESOURCE_DENIED:
             skipped.append(test)
             resource_denieds.append(test)
+        else:
```
Comment (serhiy-storchaka) [inline Lib/test/regrtest.py:483]: 3.x doesn't have this branch.

File: Lib/test/regrtest.py [hunk 2]
```diff
@@ -533,9 +538,17 @@ def work():
                                    universal_newlines=True,
                                    close_fds=(os.name != 'nt'))
                     stdout, stderr = popen.communicate()
+                    retcode = popen.wait()
+
                     # Strip last refcount output line if it exists, since it
                     # comes from the shutdown of the interpreter in the subcommand.
                     stderr = debug_output_pat.sub("", stderr)
+
+                    if retcode != 0:
```
Comment (serhiy-storchaka) [inline Lib/test/regrtest.py:547]: In 3.x this is after `stdout, _, result = stdout.strip().rpartition("\n")`.

File: Lib/test/regrtest.py [hunk 3]
```diff
@@ -533,9 +538,17 @@ def work():
...
+                        result = (CHILD_ERROR, "Exit code %s" % retcode)
+                        output.put((test, stdout.rstrip(), stderr.rstrip(),
+                                    result))
+
```
Comment (serhiy-storchaka) [inline Lib/test/regrtest.py:551]: Missed `return`?

File: Lib/test/regrtest.py [hunk 4]
```diff
@@ -545,9 +558,11 @@ def work():
             except BaseException:
                 output.put((None, None, None, None))
                 raise
+
```
Comment (serhiy-storchaka) [inline Lib/test/regrtest.py:561]: There are no empty lines in 3.5.

File: Lib/test/regrtest.py [hunk 5]
```diff
@@ -556,21 +571,24 @@ def work():
...
                 if result[0] == INTERRUPTED:
                     assert result[1] == 'KeyboardInterrupt'
                     raise KeyboardInterrupt   # What else?
-                accumulate_result(test, result)
-                if not quiet:
-                    fmt = "[{1:{0}}{2}/{3}] {4}" if bad else "[{1:{0}}{2}] {4}"
-                    print(fmt.format(
-                        test_count_width, test_index, test_count,
-                        len(bad), test))
+
```
Comment (serhiy-storchaka) [inline Lib/test/regrtest.py:591]: In 3.5:
```
if result[0] == CHILD_ERROR:
    raise Exception("Child error on {}: {}".format(test, result[1]))
```
In master this is handled in different way.

File: Lib/test/regrtest.py (hunk 1 above)
Comment (vstinner) [reply Lib/test/regrtest.py:483]: It's on purpose: it should catch bugs in regrtest itself :-D

File: Lib/test/regrtest.py (hunk 2 above)
Comment (vstinner) [reply Lib/test/regrtest.py:547]: Oh... it's a bug in master. If the process crashed, we must not modify stdout but dump it unchanged!

File: Lib/test/regrtest.py (hunk 3 above)
Comment (vstinner) [reply Lib/test/regrtest.py:551]: Oops, it means that the worker thread will end. That's a bug :-/

File: Lib/test/regrtest.py (hunk 4 above)
Comment (vstinner) [reply Lib/test/regrtest.py:561]: I added empty lines for readability, is that an issue?

File: Lib/test/regrtest.py (hunk 5 above)
Comment (vstinner) [reply Lib/test/regrtest.py:591]: I tried to write the smallest patch to handle child error. regrtest in master is *very* different.

Comment (serhiy-storchaka) [reply Lib/test/regrtest.py:591]: That is why I compared the code with 3.5.

I don't suggest backporting the code from master (with showing the progress of parallel long running test etc), but shouldn't these lines be ported from 3.5?

File: Lib/test/regrtest.py (hunk 4 above)
Comment (serhiy-storchaka) [reply Lib/test/regrtest.py:561]: Only increasing difference between 2.7 and 3.5.

File: Lib/test/regrtest.py (hunk 1 above)
Comment (serhiy-storchaka) [reply Lib/test/regrtest.py:483]: Should this be added in 3.x?

**Pull request merged**